Message format (#3)
unicorn-wg/oblivious-http: draft-thomson-http-binary-message.md

# Introduction

This document defines a simple format for representing an HTTP message
({{!HTTP=I-D.ietf-httpbis-semantics}}), either request or response. This allows
for the encoding of HTTP messages that can be conveyed outside of an HTTP
protocol. This enables the application of transformations to entire messages,

> including the application of encryption and authentication.

Reviewer: ```suggestion
including the application of authenticated encryption.
```

... to not encourage folks to authenticate without encrypting. :)

This format is informed by the structure of HTTP/2 ({{?H2=RFC7540}}) and HTTP/3
({{?H3=I-D.ietf-quic-http}}), but omits header compression ({{?HPACK=RFC7541}},
{{?QPACK=I-D.ietf-quic-qpack}}) and does not include a generic framing layer.

This provides an alternative to the `message/http` content type defined in
{{?MESSAGING=I-D.ietf-httpbis-messaging}}.  A binary format permits more efficient encoding and processing of content.  A binary format also reduces exposure to security problems related to processing of HTTP

# Conventions and Definitions

{::boilerplate bcp14}

> This document uses terminology from HTTP ({{!HTTP}}).
> 
> This document uses notation from QUIC ({{!QUIC=I-D.ietf-quic-transport}}).

Reviewer: ```suggestion
This document uses terminology from HTTP ({{!HTTP}}) and notation from QUIC ({{!QUIC=I-D.ietf-quic-transport}}).
```

length prefix, in the order listed.

The format of request control data is shown in {{format-request-control-data}}.

~~~
Request Control Data {
  Method Length (i),
  Method (..),
  Scheme Length (i),
  Scheme (..),
  Authority Length (i),
  Authority (..),
  Path Length (i),
  Path (..),
}
~~~
{: #format-request-control-data title="Format of Request Control Data"}

## Response Control Data

The control data for a request message includes a single field that corresponds
to the `:status` pseudo-header field in HTTP/2 {{!H2}}. This field is encoded
as a single variable length integer, not a decimal string.

The format of final response control data is shown in
{{format-response-control-data}}.

~~~
Final Response Control Data {
  Status Code (i) = 200..599,
}
~~~
{: #format-response-control-data title="Format of Final Response Control Data"}

### Informational Status Codes {#informational}

This format supports informational status codes (see Section 15.2 of
{{!HTTP}}). Responses that include information status codes are encoded by
repeating the response control data and associated header field section until

> the a final status code is encoded.

Reviewer: ```suggestion
the final status code is encoded.
```

# Introduction

This document defines a simple format for representing an HTTP message
({{!HTTP=I-D.ietf-httpbis-semantics}}), either request or response. This allows
for the encoding of HTTP messages that can be conveyed outside of an HTTP

> protocol. This enables the application of transformations to entire messages,

Reviewer 2: ```suggestion
protocol. This enables the transformation of entire messages,
```

including the application of encryption and authentication.

This format is informed by the framing structure of HTTP/2 ({{?H2=RFC7540}})
and HTTP/3 ({{?H3=I-D.ietf-quic-http}}). In comparison, this format simpler by
virtue of not including either header compression ({{?HPACK=RFC7541}},
{{?QPACK=I-D.ietf-quic-qpack}}) or a generic framing layer.

This format provides an alternative to the `message/http` content type defined
in {{?MESSAGING=I-D.ietf-httpbis-messaging}}. A binary format permits more
efficient encoding and processing of messages. A binary format also reduces
exposure to security problems related to processing of HTTP messages.

Two modes for encoding are described:

* a known-length encoding includes length prefixes for all major message
  components; and

* an indefinite-length encoding enables efficient generation of messages where

>   lengths are not known when starting an encoding.

Reviewer 2: ```suggestion
  lengths are not known when encoding starts.
```

This document uses notation from QUIC ({{!QUIC=I-D.ietf-quic-transport}}).

# Format

An HTTP message is split into five sections, following the structure defined in
Section 6 of {{!HTTP}}:

1. Framing indicator. This format uses a single integer to describe framing, which describes
   whether the message is a request or response and how subsequent sections are
   formatted; see {{framing}}

2. Control data. For a request, this contains the request method and target.
   For a response, this contains the status code.

3. Header section.  This contains zero or more header fields.

4. Content.  This is a sequence of zero or more bytes.

5. Trailer section.  This contains zero or more trailer fields.

A response message can include multiple informational responses, which causes
the control data and header section to be repeated.

All lengths and numeric values are encoded using the variable-length integer
encoding from {{!QUIC}}.

## Known Length Messages

A message that has a known length at the time of construction uses the
format shown in {{format-known-length}}.

~~~ quic-format
Message with Known-Length {
  Framing (i) = 0..1,

>   Known-Length Informational Response (..) ...,

Reviewer 2: What's the meaning of the extra `...` on this line? That it is optional?

Reviewer: It means the field is repeated one or more times (I stumbled across this myself).

Author: zero or more, yeah.

Terminator field that ends a field section or sequence of content chunks. A
message that is truncated at any other point is invalid; see {{invalid}}.

Indeterminate-length messages use the same encoding for field lines as
known-length messages; see {{fields}}.

## Framing Indicator {#framing}

The start of each is a framing indicator that is a single integer that
describes the structure of the subsequent sections. The framing indicator can
take just four values:

* A value of 0 describes a request of known length.
* A value of 1 describes a response of known length.
* A value of 2 describes a request of indeterminate length.
* A value of 3 describes a response of indeterminate length.

Other values cause the message to be invalid; see {{invalid}}.

## Request Control Data

The control data for a request message includes four values that correspond to
the values of the `:method`, `:scheme`, `:authority`, and `:path` pseudo-header
fields described in HTTP/2 (Section 8.1.2.3 of {{!H2}}). These fields are
encoded, each with a length prefix, in the order listed.

The rules in Section 8.1.2.3 of {{!H2}} for constructing pseudo-header fields

> apply to the construction of these values. However, where tge `:authority`

Reviewer 2: ```suggestion
apply to the construction of these values. However, where the `:authority`
```

> pseudo-header field might be omitted in HTTP/2, a zero-length value is encoded

Reviewer 2: I like this solution 👍

repeating the response control data and associated header section until the
final status code is encoded.

The format of the informational response control data is shown in
{{format-informational}}.

~~~ quic-format
Informational Response Control Data {
  Status Code (i) = 100..199,
}
~~~
{: #format-informational title="Format of Informational Response Control Data"}

A response message can include any number of informational responses. If the
response control data includes an informational status code (that is, a value
between 100 and 199 inclusive), the control data is followed by a header
section (encoded with known- or indeterminate- length according to the framing
indicator). After the header section, another response control data block
follows.

## Header and Trailer Field Lines {#fields}

Header and trailer sections consist of zero or more field lines; see Section 5
of {{!HTTP}}. The format of a field section depends on whether the message is
known- or intermediate-length.

Each field line includes a name and a value. Both the name and value are
non-zero length sequences of bytes. The format of a field line is shown in
{{format-field-line}}.

~~~ quic-format

> Field Line {

Reviewer 2: Do we see a way forward to have a binary encoding of structured fields? While we don't need to specify that here, it'd be nice to leave it open.

Reviewer: Does anything need to change to accommodate them here? I figured they'd just be values with a different encoding.

Author: A new format is what I was thinking, or maybe use the next bit in the framing indicator.  The latter has some risks though.
